shim/src/repair/functionConstructors.js

```diff
@@ -0,0 +1,68 @@
+// Adapted from SES/Caja - Copyright (C) 2011 Google Inc.
+// https://github.com/google/caja/blob/master/src/com/google/caja/ses/startSES.js
+// https://github.com/google/caja/blob/master/src/com/google/caja/ses/repairES5.js
+
+import verifyStrictFunctionBody from '../evaluate/verifyStrictFunctionBody';
+import { getPrototypeOf, setPrototypeOf, defineProperty } from '../commons';
+
+// This block replaces the original Function, %GeneratorFunction% and
+// %AsycnGeneratorFunction%, with safe implementations that preserve SES
+// confinement. After this block is done, the originals should no longer be
+// reachable.
+
+export default function repairFunctionConstructors(confinedWindow) {
+
+    const { eval: secureEval } = confinedWindow;
+
+    /**
+     * A safe form of the {@code Function} constructor, which
+     * constructs strict functions that can only refer freely to the
+     * {@code sharedImports}.
+     *
+     * The returned function is strict whether or not it declares
+     * itself to be.
+     */
+    function secureFunctionFactory(funcDeclaration) {
+        return (function SecureFunction(...params) {
+            var body = verifyStrictFunctionBody(params.pop() || '');
```

Comment:
oh, I understand now. This is reused for function and function*, and any other form of function declaration. got it.